### ISSUE TITLE
adriconf fails without XDG_SESSION_TYPE=x11

### ISSUE THREAD

**Issue by orbea** [jlHertel/adriconf#62]
OS: `Slackware64-current`
adriconf: `1.4`
mesa: https://github.com/mesa3d/mesa/commit/d7eb40962bbf62b561cd51be94257786b2d2bbcc
xorg: https://gitlab.freedesktop.org/xorg/xserver/commit/ce9455b5ee389b100a9b7da76b79690d97211b7a
gtk+: `3.24.10`

I am using amdgpu / radeonsi.

When I start `adriconf` without ` XDG_SESSION_TYPE=x11` it fails to start with this error.
```
$ adriconf 
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
Aborted
```
When I set `XDG_SESSION_TYPE=x11` I recieve this output, I am not sure if the warning is important?
```
$ XDG_SESSION_TYPE=x11 adriconf 
adriconf running on X11
Current language code is en
Exception caught: Could not create parser context
```
This seems similar to issue https://github.com/jlHertel/adriconf/issues/52, but I have not found a satisfactory solution there excluding the environment variable.

Some things I have noticed:

* The mesa meson build does pass `-DUSE_DRICONF`.
* I do not have `/etc/drirc`, but I do have `/usr/share/drirc.d/00-mesa-defaults.conf`
* `$XDG_SESSION_TYPE` is unset by default and I use a minimal tiling window manager.

Is there a better way to have it automatically detect the correct session type?

**Comment by orbea** [comment]
This issue also seems related, but I am using xorg and not wayland.

https://github.com/jlHertel/adriconf/issues/55

**Comment by jlHertel** [comment]
Can you please try latest master code?
There are some fixes there to handle such cases.

**Comment by orbea** [comment]
I tried commit c37d155f0e43a143a83a7fab79d4c20b91bca667.

```
$ ./adriconf 
Generic Error: basic_string::_M_construct null not valid
```
```
$ XDG_SESSION_TYPE=x11 ./adriconf 
adriconf running on X11
Current language code is en
Found config under: /usr/share/drirc.d/00-mesa-defaults.conf
```

**Comment by jlHertel** [comment]
I'm experimenting with a new way to log and trace issues.
It's not totally stable, but maybe is worth a try in your case.
Can you checkout https://github.com/jlHertel/adriconf/commit/e02b8edb5a3d98e0510c49e9655641ea962cd672 from branch `improve_wayland_detection`?
After it, compile and execute using this command: `VERBOSITY=DEBUG ./adriconf`

This should give a lot more output and hopefully, and insight on why it's failing for you.

**Comment by orbea** [comment]
I checked out the `improve_wayland_detection` branch and commit e02b8edb5a3d98e0510c49e9655641ea962cd672.

However I ran into an unrelated issue.
```
-- Found GTest: /usr/lib64/libgtest.so  
-- Checking for module 'gmock'
--   Package 'gmock', required by 'virtual:world', not found
CMake Error at /usr/share/cmake-3.15/Modules/FindPkgConfig.cmake:458 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.15/Modules/FindPkgConfig.cmake:637 (_pkg_check_modules_internal)
  tests/CMakeLists.txt:26 (pkg_check_modules)


-- Configuring incomplete, errors occurred!
See also "/tmp/adriconf/build/CMakeFiles/CMakeOutput.log".
See also "/tmp/adriconf/build/CMakeFiles/CMakeError.log".
```
I could not find a gmock or googlemock package for Slackware so I looked upstream where the repo has been archived.

https://github.com/google/googlemock

> This project has been absorbed into the GoogleTest project.

I do have gtest installed, checked my gtest package (`1.8.0`) and found it includes these files.
```
usr/lib64/libgmock.so
usr/lib64/libgmock_main.so
usr/lib64/libgtest.so
usr/lib64/libgtest_main.so
```
However the error indicates that gmock is not being found, I then tried to update to `gtest-1.8.1` and found no difference.

**Comment by orbea** [comment]
I worked around this issue by removing this line from `CMakeLists.txt`.
```
add_subdirectory(tests)
```
```
$ VERBOSITY=DEBUG ./adriconf
DEBUG: Creating the GTK application object
DEBUG: Checking if the system is supported
ERROR: Generic Error: basic_string::_M_construct null not valid
```

**Comment by jlHertel** [comment]
Can you try running with this command?
`VERBOSITY=DEBUG XDG_SESSION_TYPE=x11 ./adriconf`

**Comment by orbea** [comment]
I attached a log since its more verbose than the last one.

[adriconf.txt](https://github.com/jlHertel/adriconf/files/3513205/adriconf.txt)

I spent some time trying to debug this with `printf` and I think its failing when on lines like this where `XDG_SESSION_TYPE` is unset and doesn't return anything.

https://github.com/jlHertel/adriconf/blob/c37d155f0e43a143a83a7fab79d4c20b91bca667/adriconf/DRIQuery.cpp#L18

**Comment by orbea** [comment]
This solution might work better?

> X11 uses the DISPLAY environment variable to find the X server. Wayland uses WAYLAND_DISPLAY. Look for the Wayland variable first. Then if you don't find it or you can't connect go on to using X11.
>
> Do not skip checking the WAYLAND_DISPLAY variable or assume Wayland is running on "wayland-0". Some people want to use nested compositors, which you would bypass. Other people may be running Wayland but want to force X11 rendering by deleting the WAYLAND_DISPLAY variable.

https://stackoverflow.com/questions/45536141/how-i-can-find-out-if-a-linux-system-uses-wayland-or-x11/45537237#45537237

**Comment by jlHertel** [comment]
@orbea I updated the branch to use this approach. Would you mind checking-out the latest code and checking if it works fine for you?

**Comment by orbea** [comment]
It seems to work now. :)
```
$ ./adriconf 
INFO: adriconf running on X11
```

**Comment by jlHertel** [comment]
cool.
I will leave this issue open for now, as I'm not ready to merge this branch yet.
I will close it as soon as the branch is merged and a new release is issued.

Also, thanks for reporting this issue and helping in figuring out what was the root cause. Your dedication is appreciated :)

**Comment by orbea** [comment]
Also thanks for fixing it so quickly!

**Comment by orbea** [comment]
I see this is fixed in `1.5.1` now so I am going to close this.
